### ISSUE TITLE
split function

### ISSUE THREAD

**Issue by asmeurer** [Quansight/ndindex#28]
`split(i0, [i1, i2, ...])` will return a list of indices `[j1, j2, ...]` such that `a[i0] = concat(a[i1][j1], a[i2][j2], ...)`.

This is needed primarily so that we can split an index into chunks, but this seems like a good generalization.

**Comment by asmeurer** [comment]
`split_slices` [here](https://github.com/Quansight/versioned-hdf5/blob/master/versioned_hdf5/slicetools.py) is a simple version of this.

**Comment by asmeurer** [comment]
I believe what's needed here is a way to project one slice onto another. `k` projecting onto `i` would give the index `j` such that `i[j] = k`, which works like an inverse of a composition https://github.com/Quansight/ndindex/issues/27 (although it is only a true inverse if `k` is a subset of `i`).

I'm not really sure what to call this operation. Let's just call it `(i^-1)[k]` for now for lack of a better notation. Given it,`split(i0, [i1, i2, ...])` becomes `[(i1^-1)[i0], (i2^-1)[i0], ...]`. 

Note that my "concat" notation above is very ambiguous. It doesn't specify how concatenation works in multiple dimensions. It is also only true that `a[i0]` would equal the concatenation if `a[i1]`, `a[i2]`, ... completely covered `a[i0]`. But I don't think that is a necessary requirement for this projection idea. 

An idea for the name might be `lift`, since we are "lifting" `k` into a slice on `i`. My main worry is that the operation is noncommutative, so the operation name and syntax should make it clear which direction things are happening in.

**Comment by asmeurer** [comment]
I also considered using `/` for this operation, but I'm not sure about it. A "right" inverse of `i[j] = k` is ambiguous. Given `j` and `k`, there are potentially any number of `i`s such that `i[j] = k`, since one can just expand `i` to include more elements than are indexed by `j`. Also, there may be no such `i`, if they have incompatible lengths for instance. But given `i` and `k`, if `k` is a subset of `i` (all elements indexed by `k` are also indexed by `i`) one can always find an index `j` such that `i[j] = k`. This is basically true by assumption, and the fact that an index `j` can index anything (some cases may require explicitly listing all indices via an integer array index). If `k` is not a subset of `i`, you can still find an index `j` such that `i[j] = k | i`, i.e., `j` selects those elements of `a[i]` that are also included in `a[k]`.

Here is a picture

Here `k` is a subset of `i`, so `a[i][j] == a[k]`

```
         +- k -+
         |     |
------- a ---------
       |         |
       +--- i ---+
         |     |
         +- j -+

```

Here `k` is not a subset, so `a[i][j] = a[k | i]`:

```
     +-- k --+
     |       |
------- a ---------
       |         |
       +--- i ---+
       |     |
       +- j -+
```

And one can also find a `l` so that `a[k][l] = a[k | i]`

```
       +- l -+
       |     |
     +-- k --+
     |       |
------- a ---------
       |         |
       +--- i ---+
```

So I'm not sure about the notation `k/i` for instance (or any other notation or function name, for that matter). It needs to be clear that you are getting `j` and not `l`. One is a recursive index of `i` and another is a recursive index of `k`. Since both make sense, there is no sort of type system to help you here.

**Comment by asmeurer** [comment]
Work started at https://github.com/Quansight/ndindex/pull/36 with the provisional name `as_subindex`.

**Comment by asmeurer** [comment]
This can be closed with the work on #36. There are still some cases that need to be implemented, but the basic idea is there now.
